### PR TITLE
add index,dat file's integrity checking

### PR DESCRIPTION
**Pull request by hxiaodon**
This pull request is created for issue [#321](https://github.com/chrislusf/seaweedfs/issues/321)

### CODE REVIEW

File: weed/storage/volume.go [hunk 1]
```diff
@@ -46,6 +46,57 @@ func loadVolumeWithoutIndex(dirname string, collection string, id VolumeId, need
 	e = v.load(false, false, needleMapKind)
 	return
 }
+func verifyIndexFileIntegrity(indexFile *os.File) (indexSize int64, err error) {
+	var fi os.FileInfo
+	if fi, err = indexFile.Stat(); err != nil {
+		return
+	} else if indexSize = fi.Size(); indexSize != 0 && indexSize%16 != 0 {
```
Comment (chrislusf) [inline weed/storage/volume.go]: 1) "else" is not needed. 

2) "indexSize!=0" is not needed.

3) How about defining a constant of NeedleIndexSize = 16 in needle_map.go file?

File: weed/storage/volume.go [hunk 2]
```diff
@@ -46,6 +46,57 @@ func loadVolumeWithoutIndex(dirname string, collection string, id VolumeId, need
...
+		}
+	}
+	return nil
+}
+func volumeDataIntegrityChecking(v *Volume, indexFile *os.File) {
+	var indexSize int64
+	var e error
+	if indexSize, e = verifyIndexFileIntegrity(indexFile); e != nil {
+		glog.V(0).Infof("verifyIndexFileIntegrity failed %s", e.Error())
+		v.readOnly = true
```
Comment (chrislusf) [inline weed/storage/volume.go]: add a return here?

File: weed/storage/volume.go [hunk 3]
```diff
@@ -46,6 +46,57 @@ func loadVolumeWithoutIndex(dirname string, collection string, id VolumeId, need
...
+}
+func verifyNeedleIntegrity(datFile *os.File, v Version, offset int64, key uint64) error {
+	if n, bodyLength, err := ReadNeedleHeader(datFile, v, offset); err != nil {
+		return fmt.Errorf("can not read needle header: %s", err.Error())
+	} else {
+		if n.Id != key {
+			return fmt.Errorf("index key %#x does not match needle's Id %#x", key, n.Id)
+		} else {
+			if err := n.ReadNeedleBody(datFile, v, offset+int64(NeedleHeaderSize), bodyLength, true); err != nil {
+				return fmt.Errorf("dat file %s's body reading failed: %s", datFile.Name(), err.Error())
```
Comment (chrislusf) [inline weed/storage/volume.go]: When printing errors, use fmt.Errorf("... %v", err) is more succinct.

File: weed/storage/volume.go [hunk 4]
```diff
@@ -46,6 +46,57 @@ func loadVolumeWithoutIndex(dirname string, collection string, id VolumeId, need
...
+	return
+}
+func verifyNeedleIntegrity(datFile *os.File, v Version, offset int64, key uint64) error {
+	if n, bodyLength, err := ReadNeedleHeader(datFile, v, offset); err != nil {
+		return fmt.Errorf("can not read needle header: %s", err.Error())
+	} else {
+		if n.Id != key {
+			return fmt.Errorf("index key %#x does not match needle's Id %#x", key, n.Id)
+		} else {
+			if err := n.ReadNeedleBody(datFile, v, offset+int64(NeedleHeaderSize), bodyLength, true); err != nil {
```
Comment (chrislusf) [inline weed/storage/volume.go]: Do not complicate this function: ReadNeedleBody()

Just add the checksum verification logic here.

File: weed/storage/volume.go [hunk 5]
```diff
@@ -46,6 +46,57 @@ func loadVolumeWithoutIndex(dirname string, collection string, id VolumeId, need
...
+	var indexSize int64
+	var e error
+	if indexSize, e = verifyIndexFileIntegrity(indexFile); e != nil {
+		glog.V(0).Infof("verifyIndexFileIntegrity failed %s", e.Error())
+		v.readOnly = true
+	}
+	var lastIdxEntry []byte
+	if lastIdxEntry, e = readIndexEntryAtOffset(indexFile, indexSize-16, v.Version()); e != nil {
+		glog.V(0).Infof("readLastIndexEntry failed %s", e.Error())
+		v.readOnly = true
```
Comment (chrislusf) [inline weed/storage/volume.go]: add a return here?

**Comment by chrislusf** [comment]
Thanks! Need some adjustments.

File: weed/storage/volume.go (hunk 1 above)
Comment (hxiaodon) [inline weed/storage/volume.go]: I agree with point 1 and point 3. 
Regarding point 2, I guess you mean I could use the returned indexSize to do the comparation work at outside function(assuming there is a empty idx and dat file which does not need this data integrity checking)

File: weed/storage/volume.go (hunk 2 above)
Comment (hxiaodon) [inline weed/storage/volume.go]: Sure, I will add a return type(error) and mark readOnly in the outside function

File: weed/storage/volume.go (hunk 3 above)
Comment (hxiaodon) [inline weed/storage/volume.go]: Got it. Will update

File: weed/storage/volume.go (hunk 4 above)
Comment (hxiaodon) [inline weed/storage/volume.go]: Then the ReadNeedleBody() need to add another return value: bodyData []byte

File: weed/storage/volume.go (hunk 1 above)
Comment (chrislusf) [inline weed/storage/volume.go]: 2) I was wrong with this. "indexSize!=0" is needed.

File: weed/storage/volume.go (hunk 4 above)
Comment (chrislusf) [inline weed/storage/volume.go]: How about using ReadData() function instead?

**Comment by hxiaodon** [comment]
Chris,please take a look at the latest comparing changes [#link](https://github.com/chrislusf/seaweedfs/compare/master...hxiaodon:master), if it's OK, I will create another pull request

**Comment by chrislusf** [comment]
No need for separate PR.

Can you use ReadData() function instead of changing ReadNeedleBody()?

**Comment by hxiaodon** [comment]
Sure, I just updated. Please help to review~

**Comment by chrislusf** [comment]
Thanks! I will merge and make some minor changes.

**Pull request merged**